### PR TITLE
Feature Flag Support

### PR DESCRIPTION
**Pull request by ahammer**
## Purpose

* Compile time flag framework for Android

First Step Only.
Support for Beta SDK Version, and Support for Folder based Feature Flags/Implementations.

Can be used for runtime checking, or for compile time splitting (based on API).

Made "CI Compatible", but will need to update CI to improve further (i.e. adding flavors instead of a config).

## Does this introduce a breaking change?

- [ ] Yes
- [ ] No

## Pull Request Type
What kind of change does this Pull Request introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Pull Request Checklist

- [ ] Public API changes
- [ ] Verified for cross-platform
- [ ] Synced with iOS, Web team
- [ ] Internal review completed
- [ ] UI Changes
  - [ ] Screen captures included
  - [ ] Tested for Light/Dark mode
  - [ ] Tested for screen rotation
  - [ ] Tested on Tablet and small screen device (5")
  - [ ] Include localization changes
- [ ] Tests
  - [ ] Memory leak analysis performed
  - [ ] Tested on API 21, 26 and latest 
  - [ ] Tested for background mode
  - [ ] Unit Tests Included
  - [ ] UI Automated Tests Included

## How to Test

* Open...
* Click on...

## What to Check
Verify that the following are valid
* ...

### CODE REVIEW

**Review by iaulakh — DISMISSED** [review]

**Review by Mohtasim — APPROVED** [review]

**Review by ShaunaSong — APPROVED** [review]

File: azure-communication-ui/build.gradle [hunk 1]
```diff
@@ -32,12 +32,40 @@ buildscript {
...
+            //
+            // USE_CALLING_SDK_BETA=true
+            //
+            // and then run ./gradlew clean
+            // This will use the beta version of the SDK
+            // and enable the beta features
+
+            // In this branch, it was already Beta version, so it'll be in the GA field here
+            // This should be set to a GA release before merge back
+            azure_calling_sdk_version_ga = '2.6.0-beta.6'
```
Comment (pavelprystinka) [inline azure-communication-ui/build.gradle:49]: If we read from local.properties anyway to find out which SDK version to use, why not just keep a single azure_calling_sdk_version variable and read from local.properties an override the azure_calling_sdk_version defined in build.gradle.
So if we add azure_calling_sdk_version to the local.propertyies, then use that value.

Comment (pavelprystinka) [inline azure-communication-ui/build.gradle:49]: If we read from local.properties anyway to find out which SDK version to use, why not just keep a single azure_calling_sdk_version variable and read from local.properties an override the azure_calling_sdk_version defined in build.gradle.
So if we add azure_calling_sdk_version to the local.propertyies, then use that value.

File: azure-communication-ui/calling/README.md [hunk 2]
```diff
@@ -46,7 +46,14 @@ repositories {
 ```
 Sync project with gradle files. (Android Studio -> File -> Sync Project With Gradle Files)
 
+### Beta Calling SDK and Features
```
Comment (pavelprystinka) [inline azure-communication-ui/calling/README.md:49]: I do not think we need to share this with Contoso.
They just can go and put any version they would like to test with in the build.gradle.
USE_BETA_CALLING_SDK does not necessary use actual lates released SDK beta.

File: azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/features/ACSFeature.kt [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+ *
+ * @author yassirb@microsoft.com
+ */
+internal interface ACSFeature {
+    /**
+     * Return feature availability
+     *
+     * @return true if available on the current release otherwise false
+     */
+    fun IsAvailable(): Boolean
```
Comment (pavelprystinka) [inline azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/features/ACSFeature.kt:17]: should it be `isAvailable` get only property?

File: azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/features/ACSFeatureFactory.kt [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+     * @param _instance feature instance
+     * @param <I> type of the feature interface
+     * @param <O> type of the feature instance
+     </O></I> */
+    private inline fun <reified I : ACSFeature> RegisterACSFeature(_instance: I) {
+        featureList!![I::class.java] = _instance
+    }
+
+    companion object {
+        private var featureList: MutableMap<Class<*>, ACSFeature>? = null
```
Comment (pavelprystinka) [inline azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/features/ACSFeatureFactory.kt:74]: Why not to make it non-nullable and give is a value` = HashMap()`. Then we will not need to use `!!`

File: azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/features/interfaces/ISupportFilesFeature.kt [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+// Licensed under the MIT License.
+
+package com.azure.android.communication.ui.calling.features.interfaces
+
+import android.content.Context
+import com.azure.android.communication.calling.CallClient
+import com.azure.android.communication.ui.calling.features.ACSFeature
+import java.io.File
+
+abstract class ISupportFilesFeature : ACSFeature {
```
Comment (pavelprystinka) [inline azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/features/interfaces/ISupportFilesFeature.kt:11]: It reminds me a C# code style with prefixing `I` for interfaces.
Should we rather use `SupportFilesFeature` for interface and `SupportFilesFeatureimpl` for class.
https://kotlinlang.org/docs/coding-conventions.html#function-names

**Review by pavelprystinka — CHANGES_REQUESTED** [review]

File: azure-communication-ui/build.gradle (hunk 1 above)
Comment (ahammer) [reply azure-communication-ui/build.gradle:49]: it's meant to be pinned to latest public release (with manual updates to the file as we need them). But down the road I can add an over-ride version in local.properties.

File: azure-communication-ui/calling/README.md (hunk 2 above)
Comment (ahammer) [reply azure-communication-ui/calling/README.md]: Removed from Readme

File: azure-communication-ui/build.gradle (hunk 1 above)
Comment (ahammer) [reply azure-communication-ui/build.gradle:49]: meant to be pinned to latest public release, aligning with Calling SDK
i.e. they always have a latest GA and latest Beta. This should generally be these values here.

We can add a version over-ride to local.properties if people can use. but not in this PR.

File: azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/features/ACSFeature.kt (hunk 3 above)
Comment (ahammer) [reply azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/features/ACSFeature.kt:17]: will update, I pulled this code direct from calling native

File: azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/features/ACSFeatureFactory.kt (hunk 4 above)
Comment (ahammer) [reply azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/features/ACSFeatureFactory.kt:74]: will update. Code was grabbed straight from native and went through the Kotlin convertor. I cleaned it up a bit and removed !!'s etc. Will be pushed soon.

File: azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/features/interfaces/ISupportFilesFeature.kt (hunk 5 above)
Comment (ahammer) [reply azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/features/interfaces/ISupportFilesFeature.kt:11]: Will update.

**Review by pavelprystinka — DISMISSED** [review]

**Review by pavelprystinka — APPROVED** [review]

**Review by iaulakh — APPROVED** [review]

**Review by ShaunaSong — APPROVED** [review]

**Pull request merged**